language/ast.py

```diff
@@ -0,0 +1,779 @@
+    """
+
+    package = Field(type=UnqualifiedID)
+    name = Field(type=UnqualifiedID)
+
+
+@abstract
+class Type(RFLXNode):
+    """
+    Base class for type declarations (enums, integers, messages, arrays, derivations)
+    """
+
+
+@abstract
+class TypeDef(RFLXNode):
+    """
+    Base class for type definitions
+    """
+
+
+class TypeSpec(Type):
+    """
+    Type specification (type Foo is ...)
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    definition = Field(type=TypeDef)
+
+
+@abstract
+class Expr(RFLXNode):
+    """
+    Base class for expressions
+    """
+
+
+class Op(RFLXNode):
+    """
+    Operators for expressions
+    """
+
+    enum_node = True
+    alternatives = [
+        "pow",
+        "mul",
+        "div",
+        "add",
+        "sub",
+        "mod",
+        "eq",
+        "neq",
+        "le",
+        "lt",
+        "gt",
+        "ge",
+        "and",
+        "or",
+        "in",
+        "notin",
+    ]
+
+
+class Negation(Expr):
+    """
+    Negation
+    """
+
+    data = Field(type=Expr)
+
+
+class BinOp(Expr):
+    """
+    Binary operation
+    """
+
+    left = Field(type=Expr)
+    op = Field(type=Op)
+    right = Field(type=Expr)
+
+
+class ParenExpression(Expr):
+    """
+    Parenthesized expression
+    """
+
+    data = Field(type=Expr)
+
+
+class RefinementSpec(Type):
+    """
+    Refinement specification (for Message use (Field => Inner_Type))
+    """
+
+    pdu = Field(type=ID)
+    field = Field(type=UnqualifiedID)
+    sdu = Field(type=ID)
+    condition = Field(type=Expr)
+
+
+@abstract
+class TypeDecl(RFLXNode):
+    """
+    Base class for type declarations
+    """
+
+
+class PrivateTypeDecl(TypeDecl):
+    """
+    Private session type declaration
+    """
+
+    identifier = Field(type=UnqualifiedID)
+
+
+class SessionAspects(RFLXNode):
+    """
+    Session aspects (Initial, Final)
+    """
+
+    initial = Field(type=UnqualifiedID)
+    final = Field(type=UnqualifiedID)
+
+
+@abstract
+class SessionDecl(RFLXNode):
+    """
+    Base class for session declarations
+    """
+
+
+class VariableDecl(SessionDecl):
+    """
+    Session variable declaration
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    type_identifier = Field(type=ID)
+    initializer = Field(type=Expr)
+
+
+class RenamingDecl(SessionDecl):
+    """
+    Session renaming declaration
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    type_identifier = Field(type=ID)
+    expression = Field(type=Expr)
+
+
+@abstract
+class BaseStateBody(RFLXNode):
+    """
+    Base class for session state body
+    """
+
+
+class NullStateBody(BaseStateBody):
+    """
+    Null session state body
+    """
+
+
+@abstract
+class Statement(RFLXNode):
+    """
+    Base class for statements
+    """
+
+
+class Assignment(Statement):
+    """
+    Assignment of expression to unqualified identifier
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    expression = Field(type=Expr)
```

Comment:
Irc there are neither statements nor assignments in the message specification. Is this part of the session API? If so maybe this should be at least documented.

Reply:
This is part of the session. Have a look e.g. [here](https://github.com/Componolit/RecordFlux/tree/develop/tests/integration/session).